rfd/0000-sre-challenge.md

```diff
@@ -0,0 +1,264 @@
+### Design Document for Kubernetes Deployment Scaler
+
+---
+authors: Derek Norrbom (dereknorrbom@gmail.com)
+state: draft
+---
+
+## What
+
+This document outlines the proposed design and implementation plan for a Kubernetes Deployment Scaler API. The API will provide functionality to get and set the replica count of a Kubernetes deployment, utilizing mTLS for secure communication and caching for efficient data retrieval.
+
+## Why
+
+The purpose of this project is to implement a scalable, secure API for managing Kubernetes deployments. This project will demonstrate proficiency in Go, Kubernetes, Docker, and security best practices (mTLS). It will also highlight the ability to design, build, and document an application that can be easily extended and maintained.
+
+## How
+
+### API Structure
+
+The API will provide the following endpoints:
+
+- **GET /healthz**: Health check endpoint that returns "OK" if the server is running and Kubernetes connectivity is verified.
+- **GET /replica-count?namespace={namespace}&deployment={deployment}**: Returns the current replica count of the specified deployment.
+- **POST /set-replica-count?namespace={namespace}&deployment={deployment}**: Sets the replica count of the specified deployment.
```

Comment:
These two endpoints are very similar. They both accept the same parameters and operate on replica count. What is the reason for defining two endpoints rather than one endpoint with different HTTP verbs?

Reply:
In pedantic REST, the endpoints would use the same name with different verbs. I went with descriptive names for clarity.

Proposed Update:
- **GET /replica-count?namespace={namespace}&deployment={deployment}**: Returns the current replica count of the specified deployment.
- **POST /replica-count?namespace={namespace}&deployment={deployment}**: Sets the replica count of the specified deployment.